common/perf-test-core/src/main/java/com/azure/perf/test/core/ApiPerfTestBase.java

```diff
@@ -225,6 +229,72 @@ public Mono<Void> runAllAsync(long endNanoTime) {
             .then();
     }
 
+    @Override
+    public CompletableFuture<Void> runAllAsyncWithCompletableFuture(long endNanoTime) {
+        completedOperations = 0;
+        lastCompletionNanoTime = 0;
+        long startNanoTime = System.nanoTime();
+        Semaphore semaphore = new Semaphore(10); // Limit to 10 concurrent tasks
+
+        CompletableFuture<Void> future = CompletableFuture.completedFuture(null);
+
+        return runTasksRecursively(future, endNanoTime, startNanoTime, semaphore);
+    }
+
+    private CompletableFuture<Void> runTasksRecursively(CompletableFuture<Void> future, long endNanoTime, long startNanoTime, Semaphore semaphore) {
+        if (System.nanoTime() >= endNanoTime) {
+            return future;
+        }
+
+        return future.thenCompose(ignored -> {
+            try {
+                semaphore.acquire();
+                return runTestAsyncWithCompletableFuture()
+                    .thenAccept(result -> {
+                        completedOperations += result;
+                        lastCompletionNanoTime = System.nanoTime() - startNanoTime;
+                    })
+                    .whenComplete((res, ex) -> semaphore.release());
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                throw new RuntimeException(e);
+            }
+        }).thenCompose(ignored -> runTasksRecursively(future, endNanoTime, startNanoTime, semaphore));
+    }
+
+
+    @Override
+    public Runnable runAllAsyncWithExecutorService(long endNanoTime) {
+        completedOperations = 0;
+        lastCompletionNanoTime = 0;
+        long startNanoTime = System.nanoTime();
+
+        return () -> {
+            while (System.nanoTime() < endNanoTime) {
+                try {
+                    int result = Executors.newFixedThreadPool(1).submit(() -> runTestAsyncWithCompletableFuture().get()).get();
+                    completedOperations += result;
+                    lastCompletionNanoTime = System.nanoTime() - startNanoTime;
```

Comment:
Is this measuring the right duration? Should the `startNanoTime` be reset before each `submit()`?

Reply:
yeah updated it to reset before.